File: CtCI_graphs_trees.js
// 4.2 Minimal Tree
// Given a sorted (ascending order) array with unique integer elements,
// write an algorithm to build a binary search tree with minimal height

function BSTNode(val) {
  this.val = val;
  this.left = null;
  this.right = null;
}

function minimalBST(array) {
  if(!array.length) { return null; }
  const midpoint = Math.floor(array.length / 2);
  const rootNode = new BSTNode(array[midpoint]);

  let left = array.slice(0, midpoint);
  let right = array.slice(midpoint + 1);

  rootNode.left = minimalBST(left);
  rootNode.right = minimalBST(right);

  return rootNode;
}


// #======================================================================
// #======================================================================
//
// 4.3 List of Depths
// Given a binary tree, design an algorithm which creates a linked list of all
// the nodes at each depth (e.g. if you have a tree with depth D, you'll have
// D linked lists)

  //Use the BSTNodes defined in 4.2 above
  // Even though this is just a binary tree and not a BST

  function ListNode(val) {
    this.val = val;
    this.next = null;
  }

  function listOfDepths(rootNode) {
    const lists = {};

    const _createLists = (root, level) => {
      if(root === null) { return; }
      let newListNode = new ListNode(root.val);

      if(lists[level] !== undefined) { newListNode.next = lists[level]; }
      lists[level] = newListNode;

      level++;
      _createLists(root.left, level);
      _createLists(root.right, level);
    };

    _createLists(rootNode, 1);
    return lists;
  }

// TEST CODE
//   let nodeA = new BSTNode("a");
//   let nodeB = new BSTNode("b");
//   let nodeC = new BSTNode("c");
//   let nodeD = new BSTNode("d");
//   let nodeE = new BSTNode("e");
//   let nodeF = new BSTNode("f");
//   let nodeG = new BSTNode("g");
//   nodeA.left = nodeB;
//   nodeA.right = nodeC;
//   nodeB.left = nodeD;
//   nodeB.right = nodeE;
//   nodeC.left = nodeF;
//   nodeF.left = nodeG;
//
//   console.log(nodeA);
//   const listOfLists = listOfDepths(nodeA);
//   console.log(listOfLists);

// #======================================================================
// #======================================================================

File: CtCI_graphs_trees.rb
# 4.1 - Route between nodes
# Given a directed graph, design an algorithm to find
# out whether there is a route between two nodes

class Node
  attr_accessor :edges

  def initialize(value)
    @value = value
    @edges = []
  end
end

class Edge
  attr_reader :fromNode, :toNode
  def initialize(fromNode, toNode)
    @fromNode = fromNode
    @toNode = toNode
    fromNode.edges.push(self)
  end
end

def route?(startNode, endNode)
  visited = Hash.new
  queue = Queue.new
  queue.push(startNode)

  until queue.empty?
    currentNode = queue.pop
    return true if currentNode == endNode

    currentNode.edges.each do |edge|
      child = edge.toNode
      unless visited[child]
        queue.push(child)
        visited[child] = true
      end
    end
  end

  false
end

# test case

nodeA = Node.new(:A) #Start Node
nodeB = Node.new(:B) #End Node
nodeC = Node.new(:C)
nodeD = Node.new(:D)
nodeE = Node.new(:E)
nodeF = Node.new(:F)
nodeG = Node.new(:G)

Edge.new(nodeA, nodeF)
Edge.new(nodeF, nodeG)
Edge.new(nodeG, nodeF)
Edge.new(nodeA, nodeC)
Edge.new(nodeC, nodeD)
Edge.new(nodeD, nodeB)
Edge.new(nodeB, nodeD)
Edge.new(nodeB, nodeE)


#======================================================================
#======================================================================

 # 4.3 List of Depths
 # Given a binary tree, design an algorithm which creates a linked list of all
 # the nodes at each depth (e.g. if you have a tree with depth D, you'll have
 # D linked lists)

 


 # ======================================================================
 # ======================================================================
